mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,346 @@
+    return discr.inverse_mass(-div_r_vol + discr.face_mass(r_flux_bnd))
+
+
+def artificial_viscosity(discr, t, eos, boundaries, q, alpha, **kwargs):
+    """Interface :function:`av_operator` with backwards-compatible API."""
+    from warnings import warn
+    warn("Do not call artificial_viscosity; it is now called av_operator. This"
+         "function will disappear in 2021", DeprecationWarning, stacklevel=2)
+    return av_operator(discr=discr, eos=eos, boundaries=boundaries,
+                       q=q, alpha=alpha, t=t)
+
+
+def linear_operator_kernel():
+    """Apply linear operator to all elements."""
+    from meshmode.array_context import make_loopy_program
+
+    knl = make_loopy_program(
+        """{[iel,idof,j]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in}""",
+        "result[iel,idof] = sum(j, mat[idof, j] * vec[iel, j])",
+        name="modal_decomp",
+    )
+    knl = lp.tag_array_axes(knl, "mat", "stride:auto,stride:auto")
+    return knl
+
+
+def compute_smoothness_indicator():
+    """Compute the smoothness indicator for all elements."""
+    from meshmode.array_context import make_loopy_program
+
+    knl = make_loopy_program(
+        """{[iel,idof,j,k]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in and
+        0<=k<ndiscr_nodes_in}""",
+        "result[iel,idof] = "
+        "sum(k,vec[iel,k]*vec[iel,k]*modes[k])/sum(j,"
+        " vec[iel,j]*vec[iel,j]+1.0e-12/ndiscr_nodes_in)",
+        name="smooth_comp",
+    )
+    return knl
+
+
+def smoothness_indicator(discr, u, kappa=1.0, s0=-6.0):
+    """Calculate the smoothness indicator.
+
+    Parameters
+    ----------
+    u
+        A DOF Array of the field that is used to calculate the
+        smoothness indicator.
+
+    kappa
+        A optional argument that sets the controls the width of the
+        transition between 0 to 1.
+    s0
+        A optional argument that sets the smoothness level to limit
+        on. Logical values are [0,-infinity) where -infinity results in
+        all cells being tagged and 0 results in none.
+
+    Returns
+    -------
+    meshmode.dof_array.DOFArray
+        A DOF Array containing elementwise constant values between 0 and 1
+        which indicate the smoothness of a given element.
+    """
+    assert isinstance(u, DOFArray)
+
+    def get_kernel():
+        return linear_operator_kernel()
+
+    def get_indicator():
+        return compute_smoothness_indicator()
+
+    # Convert to modal solution representation
+    actx = u.array_context
+    uhat = discr.empty(actx, dtype=u.entry_dtype)
+    for group in discr.discr_from_dd("vol").groups:
+        vander = vandermonde(group.basis(), group.unit_nodes)
+        vanderm1 = np.linalg.inv(vander)
+        actx.call_loopy(
+            get_kernel(),
+            mat=actx.from_numpy(vanderm1),
+            result=uhat[group.index],
+            vec=u[group.index],
+        )
```

Comment:
With https://github.com/inducer/grudge/pull/77, this becomes way simpler (plus it does all the caching of the Vandermonde matrices/inverses so you don't recompute)

```python
from grudge.dof_desc import DD_MODAL, DD_VOLUME

modal_map = discr.dcoll.connection_from_dds(DD_VOLUME, DD_MODAL)
uhat = modal_map(u)
```

Reply:
Nice.

Reply:
Done, Thanks! https://github.com/illinois-ceesd/mirgecom/pull/213/commits/7dfa8143c5162dada86b5d819469fe53d42128ff